rl4co/envs/routing/cvrptw.py

```diff
@@ -0,0 +1,162 @@
+from typing import Optional
+import torch
+from tensordict.tensordict import TensorDict
+from torchrl.data import (
+    BoundedTensorSpec,
+    CompositeSpec,
+    UnboundedContinuousTensorSpec,
+    UnboundedDiscreteTensorSpec,
+)
+from zmq import device
+
+from rl4co.envs.routing.cvrp import CVRPEnv, CAPACITIES
+from rl4co.utils.ops import gather_by_index, get_distance
+
+
+class CVRPTWEnv(CVRPEnv):
+    """
+    An implementation of the Capacitated Vehicle Routing Problem (CVRP) with Time Windows (CVRPTW) environment.
+    Inherits from the CVRPEnv class.
+    """
+
+    name = "cvrptw"
+
+    def __init__(
+        self,
+        min_time: float = 0.0,
+        max_time: float = 100.0,
+        **kwargs,
+    ):
+        self.min_time = min_time
+        self.max_time = max_time
+        super().__init__(**kwargs)
+
+    def _make_spec(self, td_params: TensorDict):
+        super()._make_spec(td_params)
+
+        current_time = UnboundedContinuousTensorSpec(shape=(1), dtype=torch.float32)
+
+        durations = BoundedTensorSpec(
+            low=self.min_time,
+            high=self.max_time,
+            shape=(self.num_loc, 1),
+            dtype=torch.float32,
+        )
+
+        time_windows = BoundedTensorSpec(
+            low=self.min_time,
+            high=self.max_time,
+            shape=(
+                self.num_loc,
+                2,
+            ),  # each location has a 2D time window (start, end)
+            dtype=torch.float32,
+        )
+
+        num_vehicles = UnboundedDiscreteTensorSpec(shape=(1), dtype=torch.int64)
+        # vehicle_idx = UnboundedDiscreteTensorSpec(shape=(1), dtype=torch.int64)
+
+        # extend observation specs
+        self.observation_spec = CompositeSpec(
+            **self.observation_spec,
+            current_time=current_time,
+            durations=durations,
+            num_vehicles=num_vehicles,
+            time_windows=time_windows,
+            # vehicle_idx=vehicle_idx,
+        )
+
+    def generate_data(self, batch_size) -> TensorDict:
+        td = super().generate_data(batch_size)
+        # batch_size = [td["locs"].shape[0]]
+        batch_size = [batch_size] if isinstance(batch_size, int) else batch_size
+
+        # initialize at zero
+        current_time = torch.zeros(*batch_size, 1, dtype=torch.float32)
+        num_vehicles = torch.ones(*batch_size, dtype=torch.int64)
+
+        # define time windows
+        min_ts = torch.FloatTensor(*batch_size, self.num_loc + 1).uniform_(
+            self.min_time, self.max_time
+        )
+        max_ts = torch.FloatTensor(*batch_size, self.num_loc + 1).uniform_(
+            self.min_time, self.max_time
+        )
+        min_times, max_times = torch.min(min_ts, max_ts), torch.max(min_ts, max_ts)
+        time_windows = torch.stack((min_times, max_times), dim=-1)
+        time_windows[:, 0, :] = 0.0  # depot has no time window
```

Comment:
See comment below about ellipses `...`

Reply:
Question: if I use ellipses here, how does the machine know that it’s the second dimension I want to index here? If there could be several before and several after…?

Reply:
Right, in this case, it will be `[..., 0, :]` (machine reads: index second dimension from last.), which is a pythonic way to index on `dim=-2`